File: algorithms/L/Longest Happy String/Longest Happy String.py
class Solution:
	def longestDiverseString(self, a: int, b: int, c: int) -> str:
		pq = []
		if a > 0: heapq.heappush(pq,(-a,'a'))        
		if b > 0: heapq.heappush(pq,(-b,'b'))        
		if c > 0: heapq.heappush(pq,(-c,'c'))

		ans = ''
		while pq:
			c, ch = heapq.heappop(pq)
			if len(ans)>1 and ans[-1] == ans[-2] == ch:
				if not pq: break
				c2, ch2 = heapq.heappop(pq)
				ans += ch2
				c2 += 1
				if c2: heapq.heappush(pq,(c2,ch2))
			else:
				ans += ch
				c += 1
			if c: heapq.heappush(pq,(c,ch))

		return ans

File: scripts/algorithms/L/Longest Happy String/Longest Happy String.cpp
// Runtime: 0 ms (Top 100.0%) | Memory: 6.60 MB (Top 42.97%)

class Solution {
public:
    string longestDiverseString(int a, int b, int c) {
        //using max heap
        priority_queue<pair<int,char>>pq;
        if(a)
        pq.push({a,'a'});
        if(b)
        pq.push({b,'b'});
        if(c)
        pq.push({c,'c'});
        string ans="";
        while(pq.size()>1){
            pair<int,char>one = pq.top();pq.pop();
            pair<int,char>two = pq.top();pq.pop();
            if(one.first>=2){
                ans+=one.second;
                ans+=one.second;
                one.first-=2;
            }
            else{
                ans+=one.second;
                one.first-=1;
            }
            if(two.first>=2 && two.first>=one.first){
                ans+=two.second;
                ans+=two.second;
                two.first-=2;
            }
            else{
                ans+=two.second;
                two.first-=1;
            }
            if(one.first>0)
                pq.push(one);
            if(two.first>0)
                pq.push(two);
        }
        if(pq.empty())
            return ans;
        if(pq.top().first>=2){
            ans+=pq.top().second;
            ans+=pq.top().second;
        }
        else{
            ans+=pq.top().second;
        }
        return ans;
        
    }
};
